akd_core/src/utils.rs

```diff
@@ -33,37 +46,145 @@ pub fn get_marker_version_log2(version: u64) -> u64 {
+            || start_version_log2 != past_marker_versions[past_marker_versions.len() - 1])
+    {
+        past_marker_versions.push(start_version_log2);
+    }
+
+    let start_version_length = get_bit_length(start_version);
+    for i in (0..start_version_length).rev() {
+        let shift = 1 << i;
+        // Check if the bit of start_version at position i is 1
+        if start_version & shift != 0 {
+            let shift_mask = (shift - 1) | shift;
+            let past_version = start_version & !shift_mask;
+            if past_version != 0
+                && (past_marker_versions.is_empty()
+                    || past_version != past_marker_versions[past_marker_versions.len() - 1])
+            {
+                past_marker_versions.push(past_version);
+            }
+        }
     }
 
     // Compute future marker versions
+    let mut future_marker_versions: Vec<u64> = Vec::new();
+
+    let end_version_length = get_bit_length(end_version);
+    let mut future_version: u64 = end_version;
+    for i in 0..end_version_length {
+        let shift = 1 << i;
+        // Check if the bit of end_version at position i is 0
+        if end_version & shift == 0 {
+            future_version |= shift;
+            future_version &= !(shift - 1);
+            if future_version <= epoch {
+                future_marker_versions.push(future_version);
+            }
+        }
+    }
+
+    let endv_index: usize = find_max_index_in_skiplist(end_version);
+    let epoch_index: usize = find_max_index_in_skiplist(epoch);
+    let skiplist_slice = &MARKER_VERSION_SKIPLIST[endv_index + 1_usize..epoch_index + 1_usize];
+
     let next_marker_log2 = get_marker_version_log2(end_version) + 1;
     let final_marker_log2 = get_marker_version_log2(epoch);
-    let mut future_marker_versions: Vec<u64> =
-        ((end_version + 1)..(1 << next_marker_log2)).collect();
     for i in next_marker_log2..(final_marker_log2 + 1) {
+        let val = 1 << i;
+        if !skiplist_slice.is_empty() && val >= skiplist_slice[0] {
+            // Don't need to add any more powers of 2, can just append the skiplist slice
+            break;
+        }
         future_marker_versions.push(1 << i);
     }
+    future_marker_versions.extend_from_slice(skiplist_slice);
 
     (past_marker_versions, future_marker_versions)
 }
 
+// Given an input u64 and a sorted array of u64s, find the largest index for which
+// the corresponding array element is less than the input.
+//
+// This implementation performs a linear search over the MARKER_VERSION_SKIPLIST array,
+// but since it is sorted, it could be faster to do a binary search. However, given that
+// the array is small, there shouldn't be too much of a difference
+// between a binary search and a linear one. But if this ends up being problematic
+// in the future, it could certainly be optimized.
+//
+// Note that if the input is less than the smallest element of the array, then this
```

Comment:
Hmm, this actually will not panic either. This is another underflow area!

Reply:
Ah, thanks! Fixed in the latest update by adding explicit checks that result in panics